### PR TITLE
S1694: Promote C# to SonarWay and remove protected constructor

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to https://github.com/SonarSource/sonar-dotnet/issues/8795

## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S1694/csharp/rule.adoc [hunk 1]
```diff
@@ -2,11 +2,7 @@
...
 The purpose of an abstract class is to provide some heritable behaviors while also defining methods which must be implemented by sub-classes.
 
-
-A ``++class++`` with no abstract methods that was made ``++abstract++`` purely to prevent instantiation should be converted to a concrete ``++class++`` (i.e. remove the ``++abstract++`` keyword) with a ``++protected++`` constructor.
-
-
-A ``++class++`` with only ``++abstract++`` methods and no inheritable behavior should be converted to an ``++interface++``.
+A `class` with only `abstract` methods and no inheritable behavior should be converted to an `interface`.
 
 === Noncompliant code example
```
Comment (CristianAmbrosini) [inline rules/S1694/csharp/rule.adoc:7]: To promote it to SonarWay we should also improve a bit the documentation to LaYC standards. You can find the educational standards [here](https://github.com/SonarSource/rspec/blob/master/docs/description.adoc#3-education-format).

Comment (CristianAmbrosini) [inline rules/S1694/csharp/rule.adoc:7]:
```suggestion
=== Code examples
==== Noncompliant code example
[source,csharp,diff-id=1,diff-type=noncompliant]
...
```

**Review by CristianAmbrosini — CHANGES_REQUESTED** [review]
Let's add a How to Fix it and Resources section

File: rules/S1694/csharp/rule.adoc [hunk 2]
```diff
@@ -1,77 +1,84 @@
...
+Using interfaces over pure abstract classes presents multiple advantages:
 
+* https://en.wikipedia.org/wiki/Multiple_inheritance[**Multiple Inheritance**]: Unlike classes, an interface doesn't count towards the single inheritance limit in C#. This means a class can implement multiple interfaces, which can be useful when you need to define behavior that can be shared across multiple classes.
+* https://en.wikipedia.org/wiki/Loose_coupling#In_programming[**Loose Coupling**]: Interfaces provide a way to achieve loose coupling between classes. This is because an interface only specifies what methods a class must have, but not how they are implemented. This makes it easier to swap out implementations without changing the code that uses them.
+* https://en.wikipedia.org/wiki/Polymorphism_(computer_science)[**Polymorphism**]: Interfaces allow you to use polymorphism, which means you can use an interface type to refer to any object that implements that interface. This can be useful when you want to write code that can work with any class that implements a certain interface, _without knowing what the actual class is_.
```
Comment (CristianAmbrosini) [inline rules/S1694/csharp/rule.adoc:13]: Isn't this the same for abstract classes? [Example](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQCID2AyjWAKYIAWsA5gBQxMDuAAlrcAlCIDc1eoxbsY3XoIDCEBFzGTK5fABYh05m05cAgjCiQANgIjmrIgQF4AfDbsRLAOgCyEANZMDOgY6poU+ADMNsAAzgjIEADGCAJEAmYWHhQA3hQC+alRELHxSSm6Ar4BQTAhEhQAvhThUWnCAiDp7pY5eQWRAjQAbkzIyFAYTKl6VYE0wep9+bnkBWupxACcXABEAOo0NABmAIQ79asFTeTXLamEAiopnRlWvZf5A8Oj45PTlf45gsREsBCt1v0trtvEwaHwzhc1tcGkA===)

File: rules/S1694/csharp/rule.adoc [hunk 3]
```diff
@@ -1,77 +1,84 @@
...
+* https://en.wikipedia.org/wiki/Design_by_contract[**Design by contract**]: Interfaces provide a clear contract of what a class should do, without specifying how it should do it. This makes it easier to understand the intended behavior of a class, and to ensure that different implementations of an interface are consistent with each other.
 
-A ``++class++`` with only ``++abstract++`` methods and no inheritable behavior should be converted to an ``++interface++``.
+=== Exceptions
 
-=== Noncompliant code example
+`abstract` classes that contain non-abstract methods, in addition to `abstract` ones, cannot easily be converted to interfaces, and are not the subject of this rule:
```
Comment (CristianAmbrosini) [inline rules/S1694/csharp/rule.adoc:18]: Should we mention `fields` as well?

File: rules/S1694/csharp/rule.adoc [hunk 4]
```diff
@@ -1,77 +1,84 @@
...
 
 [source,csharp]
 ----
-public abstract class Animal // Noncompliant; should be an interface
+public abstract class Lamp // Compliant: Glow is abstract, but FlipSwitch is not
 {
-  abstract void Move();
-  abstract void Feed();
-}
+  private bool switchLamp = false;
```
Comment (CristianAmbrosini) [inline rules/S1694/csharp/rule.adoc:24]: Also this field makes it compliant

**Review by CristianAmbrosini — APPROVED** [review]
Looks good! Left a few minor comments and 1 question

File: rules/S1694/csharp/rule.adoc (hunk 3 above)
Comment (zsolt-kolbay-sonarsource) [inline rules/S1694/csharp/rule.adoc:18]: What about abstract classes that have protected abstract methods by design? If we don't want the method to be public in the derived class? Does the rule looks at this as an exception?

``` cs
public abstract class BaseClass
{
   protected abstract void SomeMethod();
}
```

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
A few questions/polishing.

File: rules/S1694/csharp/rule.adoc (hunk 3 above)
Comment (antonioaversa) [reply rules/S1694/csharp/rule.adoc:18]: Done

Comment (antonioaversa) [reply rules/S1694/csharp/rule.adoc:18]: @pavel-mikula-sonarsource

File: rules/S1694/csharp/rule.adoc (hunk 4 above)
Comment (antonioaversa) [reply rules/S1694/csharp/rule.adoc:24]: Mentioned in a dedicated paragraph on field.

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM

File: rules/S1694/csharp/rule.adoc (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply rules/S1694/csharp/rule.adoc:18]: There's no exception. Please create a FP. It should take 1 line to fix, so the repro&fix could be done right away :)

**Pull request merged**